log4j-layout-template-json/src/test/java/org/apache/logging/log4j/layout/template/json/resolver/TimestampResolverTest.java

```diff
@@ -0,0 +1,51 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one or more
+ * contributor license agreements. See the NOTICE file distributed with
+ * this work for additional information regarding copyright ownership.
+ * The ASF licenses this file to You under the Apache license, Version 2.0
+ * (the "License"); you may not use this file except in compliance with
+ * the License. You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the license for the specific language governing permissions and
+ * limitations under the license.
+ */
+package org.apache.logging.log4j.layout.template.json.resolver;
+
+import org.apache.logging.log4j.core.LogEvent;
+import org.apache.logging.log4j.layout.template.json.LogEventFixture;
+import org.apache.logging.log4j.layout.template.json.util.JsonWriter;
+import org.junit.jupiter.api.RepeatedTest;
+
+import java.util.Collections;
+
+import static org.assertj.core.api.Assertions.assertThat;
+
+class TimestampResolverTest
+{
+    private static final TemplateResolverConfig TEMPLATE_RESOLVER_CONFIG = new TemplateResolverConfig(
+            Collections.singletonMap("pattern", Collections.singletonMap("format", "yyyy-MM-dd")));
+
+    /**
+     * Tests LOG4J2-3087 race when creating layout on the same instant as the log event would result in an unquoted date in the JSON
+     */
+    @RepeatedTest( 20 )
```

Comment:
Alternatively, we could expose `TimestampResolver.FormatResolverContext#calendar` and set the same time on it as the event. This will be more predictable and would not require us to "brute-force" it.

Reply:
I would definitely favor a deterministic approach like that. This said, judging from the sources of `TimestampResolver`, code needs to be polluted a bit. Nevertheless, it is doable:

```diff
Index: log4j-layout-template-json/src/main/java/org/apache/logging/log4j/layout/template/json/resolver/TimestampResolver.java
===================================================================
diff --git a/log4j-layout-template-json/src/main/java/org/apache/logging/log4j/layout/template/json/resolver/TimestampResolver.java b/log4j-layout-template-json/src/main/java/org/apache/logging/log4j/layout/template/json/resolver/TimestampResolver.java
--- a/log4j-layout-template-json/src/main/java/org/apache/logging/log4j/layout/template/json/resolver/TimestampResolver.java	(revision a130d8ade602f36a1416eb9ef962935ea07dc4ea)
+++ b/log4j-layout-template-json/src/main/java/org/apache/logging/log4j/layout/template/json/resolver/TimestampResolver.java	(date 1620285824290)
@@ -195,6 +195,8 @@
 
     private final EventResolver internalResolver;
 
+    static Calendar CALENDAR_TEST_PROBE = null;
+
     TimestampResolver(final TemplateResolverConfig config) {
         this.internalResolver = createResolver(config);
     }
@@ -229,7 +231,7 @@
                 final FastDateFormat timestampFormat) {
             this.timestampFormat = timestampFormat;
             this.formattedTimestampBuilder = new StringBuilder();
-            this.calendar = Calendar.getInstance(timeZone, locale);
+            this.calendar = CALENDAR_TEST_PROBE = Calendar.getInstance(timeZone, locale);
             timestampFormat.format(calendar, formattedTimestampBuilder);
         }
 

```
I would appreciate it if you can go forward with this proposal of yours.

Reply:
I took a different approach instead without using a static field since that will break when running any concurrent tests.

Reply:
:man_facepalming: Yours is way better!